restore-hooks_product-requirements.md

```diff
@@ -0,0 +1,337 @@
+
+This feature will mirror the actions of a Velero backup by allowing Velero to check for any restore hooks specified for a pod. 
+
+Assumptions
+
+
+
+*   Restore operations will be run at the pod level instead of at the volume level. Some databases require the pod to be running and in some cases a user cannot manipulate a volume without the pod running. We need to support more than one type of database for this feature and so we need to ensure that this works broadly as opposed to providing support only for specific dbs. 
+*   The author of the restore hook will be responsible for invoking the restore hook. 
+
+ 
+
+MVP Use Cases 
+
+The following use cases must be included as part of the Velero restore hooks MVP (minimum viable product). 
+
+**Note: **Processing of concurrent vs sequential workloads is slated later in the Velero roadmap (see [https://github.com/vmware-tanzu/velero/pull/2548/files](https://github.com/vmware-tanzu/velero/pull/2548/files)). The MVP for this feature set will align with restore of single workloads vs concurrent workload restores. A second epic will be created to address the concurrent restore operations and will be added to the backlog for priority visibility. 
+
+**Note: **Please refer to the Requirements section of this document for more details on what items are P0 (must have in the MVP), P1 (should not ship without for the MVP), P2 (nice to haves).
+
+**<span style="text-decoration:underline;">USE CASE 1</span>**
+
+
+    **Title:** Run restore hook before pod restart.
+
+
+    **Description:  **As a user, I would like to run a restore hook before the applications in my pod are restarted.
+
+**<span style="text-decoration:underline;">______________________________________________________________</span>**
+
+**<span style="text-decoration:underline;">USE CASE 2</span>**
+
+
+    **Title: **Allow restore hook to run on non-kubernetes databases
+
+
+    **Description: **As a user, I would like to run restore hook operations even on databases that are external to kubernetes (such as postgres, elastic, etc…).
+
+**<span style="text-decoration:underline;">______________________________________________________________</span>**
+
+**<span style="text-decoration:underline;">USE CASE 3</span>**
+
+
+    **Title: **Run restore at pod level.
+
+
+    **Description: **As a user, I would like to make sure that I can run the restore hook at the pod level. And, I would like the option to run this with an annotation flag in line or using an init container.**<span style="text-decoration:underline;"> </span>**The restore pre-hook should allow the user to run the command on the container where the pre-hook should be executed. Similar to the backup hooks, this hook should run to default to fun on the first container in the pod. 
```

Comment:
Also, is this use case to allow users to choose the container in which the hook command is to be run in?

Reply:
The concept of pre- and post- hooks is called out in the use case. The design document should specify the 'before what' the pre-hook should be run against and the 'after what' the post hook should be run against.

Generally the pre-hook should be run for container, command, on-error, timeout. The post hook should be run for container, command, on-error, timeout. 

Specifications of the handling that action are not included as part of the product use case document because they should be included as part of the design document. 

The use case requires that the hook default to the run on the first container in the pod. The user should have the option to annotate/specify a different container if they do not wish to default to the first container in the pod. See the next use case for that requirement.